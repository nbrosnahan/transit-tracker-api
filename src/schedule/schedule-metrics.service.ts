import { Injectable } from "@nestjs/common"
import { MetricService } from "nestjs-otel"
import { FeedService } from "src/modules/feed/feed.service"
import { ScheduleOptions } from "./schedule.service"

interface RouteStopMetric {
  feedCode: string
  routeId: string
  stopId: string
  count: number
}

@Injectable()
export class ScheduleMetricsService {
  private readonly subscribers: Set<ScheduleOptions> = new Set()
  private readonly subscribersByFeedCode: Map<string, number> = new Map()
  private readonly routeStopMetrics: Map<string, RouteStopMetric> = new Map()

  constructor(feedService: FeedService, metricService: MetricService) {
    this.subscribersByFeedCode = new Map<string, number>(
      Object.keys(feedService.getAllFeeds()).map((feed) => [feed, 0]),
    )

    metricService
      .getObservableGauge("schedule_subscriptions", {
        description: "Number of active schedule subscriptions per feed",
        unit: "subscriptions",
      })
      .addCallback((observable) => {
        for (const [feedCode, count] of this.subscribersByFeedCode) {
          observable.observe(count, { feed_code: feedCode })
        }
      })

    metricService
      .getObservableGauge("route_stop_subscriptions", {
        description: "Number of active subscriptions for a route-stop pair",
        unit: "subscriptions",
      })
      .addCallback((observable) => {
        for (const [key, value] of this.routeStopMetrics) {
          observable.observe(value.count, {
            feed_code: value.feedCode,
            route_id: value.routeId,
            stop_id: value.stopId,
          })

          if (value.count === 0) {
            this.routeStopMetrics.delete(key)
          }
        }
      })
  }

  add(subscription: ScheduleOptions) {
    this.subscribers.add(subscription)
    this.incrementRouteStopMetrics(1, subscription)
    this.incrementFeedCodeMetrics(1, subscription)
  }

  remove(subscription: ScheduleOptions) {
    if (!this.subscribers.has(subscription)) {
      return
    }

    this.subscribers.delete(subscription)
    this.incrementRouteStopMetrics(-1, subscription)
    this.incrementFeedCodeMetrics(-1, subscription)
  }

  private incrementFeedCodeMetrics(
    value: number,
    subscription: ScheduleOptions,
  ) {
    const feedCodes: Set<string> = new Set()

    if (subscription.feedCode) {
      feedCodes.add(subscription.feedCode)
    } else {
      for (const { routeId } of subscription.routes) {
        const [feedCode] = routeId.split(":")
        feedCodes.add(feedCode)
      }
    }

    for (const feedCode of feedCodes) {
      const currentCount = this.subscribersByFeedCode.get(feedCode) ?? 0
      this.subscribersByFeedCode.set(
        feedCode,
        Math.max(0, currentCount + value),
      )
    }
  }

  private incrementRouteStopMetrics(
    value: number,
    { feedCode, routes }: ScheduleOptions,
  ) {
    feedCode = feedCode ?? "all"
    for (const routeStopPair of routes) {
      const key = `${feedCode}:${routeStopPair.routeId}:${routeStopPair.stopId}`
      const metric = this.routeStopMetrics.get(key) ?? {
        feedCode,
        routeId: routeStopPair.routeId,
        stopId: routeStopPair.stopId,
        count: 0,
      }

      metric.count += value
      this.routeStopMetrics.set(key, metric)
    }
  }
}
