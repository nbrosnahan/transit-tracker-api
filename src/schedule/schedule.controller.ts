import {
  BadRequestException,
  Controller,
  Get,
  Param,
  Query,
} from "@nestjs/common"
import {
  ApiBadRequestResponse,
  ApiOkResponse,
  ApiParam,
  ApiProperty,
} from "@nestjs/swagger"
import { ScheduleService } from "./schedule.service"

export class Trip {
  @ApiProperty({
    required: true,
    description:
      "The ID of the trip (note this is not globally unique and can be repeated for multiple days)",
    example: "1_123456",
  })
  tripId: string

  @ApiProperty({
    required: true,
    description: "The ID of the route",
    example: "1_123456",
  })
  routeId: string

  @ApiProperty({
    required: true,
    description: "The name of the route",
    example: "221",
  })
  routeName: string

  @ApiProperty({
    required: true,
    nullable: true,
    description: "The color of the route",
    example: "FF0000",
  })
  routeColor: string

  @ApiProperty({
    required: true,
    description: "The ID of the stop",
    example: "1_123456",
  })
  stopId: string

  @ApiProperty({
    required: true,
    description: "The name of the stop",
    example: "Main St & 1st Ave",
  })
  stopName: string

  @ApiProperty({
    required: true,
    description: "The headsign for the trip",
    example: "Downtown",
  })
  headsign: string

  @ApiProperty({
    required: true,
    description: "The arrival time of the trip at the stop in Unix time",
    example: 1619029200,
  })
  arrivalTime: number

  @ApiProperty({
    required: true,
    description: "The departure time of the trip at the stop in Unix time",
    example: 1619029200,
  })
  departureTime: number

  @ApiProperty({
    required: true,
    description: "A human-readable countdown text for the arrival time",
    example: "5min",
  })
  countdownText: string

  @ApiProperty({
    required: true,
    description:
      "Whether the arrival and departure times are derived from real-time data or from the static schedule",
    example: true,
  })
  isRealtime: boolean
}

export class Trips {
  @ApiProperty({
    isArray: true,
    type: Trip,
    required: true,
  })
  trips: Trip[]
}

@Controller("schedule")
export class ScheduleController {
  constructor(private readonly scheduleService: ScheduleService) {}

  private getCountdownText(arrivalTime: Date): string {
    const now = new Date()
    const diff = arrivalTime.getTime() - now.getTime()
    const hours = Math.floor(diff / 3600000)
    const minutes = Math.floor(diff / 60000)
    const seconds = Math.floor(diff / 1000)

    if (hours > 0) {
      return `${hours}h${minutes % 60}m`
    }

    if (minutes === 0 && seconds <= 30) {
      return "Now"
    }

    return `${minutes}min`
  }

  @Get(":routeStopPairs")
  @ApiOkResponse({
    description: "List of upcoming trips",
    type: Trips,
  })
  @ApiBadRequestResponse()
  @ApiParam({
    name: "routeStopPairs",
    description: "A semicolon-separated list of routeId,stopId pairs to query",
    example: "1_100113,1_71971;1_102704,1_71971",
  })
  @ApiParam({
    name: "limit",
    description: "The maximum number of trips to return",
    example: 5,
    required: false,
  })
  async getArrivals(
    @Param("routeStopPairs") routeStopPairsRaw: string,
    @Query("limit") limit: number = 10,
  ): Promise<Trips> {
    limit = Math.min(limit, 10)

    const routeStopPairs =
      this.scheduleService.parseRouteStopPairs(routeStopPairsRaw)

    if (routeStopPairs.length > 5) {
      throw new BadRequestException("Too many route-stop pairs; maximum 5")
    }

    const schedule = await this.scheduleService.getSchedule({
      routes: routeStopPairs,
      limit,
    })

    const tripDtos: Trip[] = schedule.trips
      .map((trip) => ({
        ...trip,
        countdownText: this.getCountdownText(new Date(trip.arrivalTime * 1000)),
      }))
      .sort((a, b) => a.arrivalTime - b.arrivalTime)
      .splice(0, limit)

    return {
      trips: tripDtos,
    }
  }
}
