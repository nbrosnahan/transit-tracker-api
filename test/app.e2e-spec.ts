import { INestApplication } from "@nestjs/common"
import { WsAdapter } from "@nestjs/platform-ws"
import { Test } from "@nestjs/testing"
import { StartedPostgreSqlContainer } from "@testcontainers/postgresql"
import { RedisContainer, StartedRedisContainer } from "@testcontainers/redis"
import fs from "fs/promises"
import { transit_realtime as GtfsRt } from "gtfs-realtime-bindings"
import path from "path"
import { AppModule } from "src/app.module"
import { FeedSyncService } from "src/modules/feed/feed-sync.service"
import { TripDto } from "src/schedule/schedule.controller"
import request from "supertest"
import { promisify } from "util"
import { setupFakeGtfsServer } from "./helpers/gtfs-server"
import { setupTestDatabase } from "./helpers/postgres"

describe("E2E test", () => {
  let postgresContainer: StartedPostgreSqlContainer
  let redisContainer: StartedRedisContainer
  let fakeGtfs: Awaited<ReturnType<typeof setupFakeGtfsServer>>
  let app: INestApplication

  beforeAll(async () => {
    const { postgresContainer: pgContainer, connectionUrl } =
      await setupTestDatabase()

    postgresContainer = pgContainer
    process.env.DATABASE_URL = connectionUrl.toString()

    redisContainer = await new RedisContainer().start()
    process.env.REDIS_URL = redisContainer.getConnectionUrl()

    process.env.FEEDS_CONFIG = await fs.readFile(
      path.join(__dirname, "fixtures", "feeds.test.yaml"),
      "utf-8",
    )

    fakeGtfs = await setupFakeGtfsServer()

    process.env.DISABLE_RATE_LIMITS = "true"

    const moduleRef = await Test.createTestingModule({
      imports: [AppModule],
    }).compile()

    app = moduleRef.createNestApplication()
    app.useWebSocketAdapter(new WsAdapter(app))
    await app.init()

    await app.get(FeedSyncService).syncAllFeeds()
  }, 120_000)

  afterAll(async () => {
    await app.close()
    await Promise.all([
      postgresContainer.stop(),
      redisContainer.stop(),
      promisify(fakeGtfs.server.close).bind(fakeGtfs.server)(),
    ])
  })

  test("GET /feeds", async () => {
    const response = await request(app.getHttpServer())
      .get("/feeds")
      .expect("Content-Type", /json/)
      .expect(200)

    expect(response.body).toHaveLength(1)

    const feed = response.body[0]

    expect(feed.lastSyncedAt).toBeDefined()

    const now = new Date().getTime()
    const lastSyncedAt = new Date(feed.lastSyncedAt).getTime()
    expect(lastSyncedAt).toBeLessThanOrEqual(now)
    expect(lastSyncedAt).toBeGreaterThanOrEqual(
      now - 300_000, // 5 minutes
    )

    expect(feed.code).toBe("testfeed")
    expect(feed.name).toBe("Test Feed")
    expect(feed.description).toBe("Test Feed Description")
    expect(feed.bounds).toEqual([-117.13316, 36.42529, -116.40094, 36.915684])
  })

  test("GET /feeds/service-areas", async () => {
    const response = await request(app.getHttpServer())
      .get("/feeds/service-areas")
      .expect("Content-Type", /json/)
      .expect(200)

    expect(response.body).toMatchSnapshot()
  })

  test("GET /stops/within/:bbox", async () => {
    const response = await request(app.getHttpServer())
      .get("/stops/within/-116.774095,36.909629,-116.760877,36.917066")
      .expect("Content-Type", /json/)
      .expect(200)

    expect(response.body).toHaveLength(2)
    expect(response.body).toMatchSnapshot()
  })

  test("GET /stops/:id/routes", async () => {
    const response = await request(app.getHttpServer())
      .get("/stops/testfeed:AMV/routes")
      .expect("Content-Type", /json/)
      .expect(200)

    expect(response.body).toHaveLength(1)
    expect(response.body).toMatchSnapshot()
  })

  describe("GET /schedule/:routeStopPairs", () => {
    let dateSpy: jest.SpyInstance<number, [], any>

    beforeEach(() => {
      dateSpy = jest.spyOn(Date, "now")
      dateSpy.mockImplementation(() =>
        new Date("2008-01-04T13:30:00Z").getTime(),
      )
    })

    afterEach(() => {
      dateSpy.mockRestore()
    })

    async function getTripSchedule(
      scheduleString: string = "testfeed:AAMV,testfeed:BEATTY_AIRPORT;testfeed:STBA,testfeed:STAGECOACH",
    ) {
      const response = await request(app.getHttpServer())
        .get(`/schedule/${scheduleString}`)
        .expect("Content-Type", /json/)
        .expect(200)

      expect(response.body).toHaveProperty("trips")
      return response.body.trips as TripDto[]
    }

    test("with static schedule", async () => {
      const trips = await getTripSchedule()
      expect(trips).toMatchSnapshot()
    })

    test("with service exception in static schedule", async () => {
      dateSpy.mockImplementation(() =>
        new Date("2007-06-04T13:30:00Z").getTime(),
      )

      const trips = await getTripSchedule()
      expect(trips).toMatchSnapshot()

      // Expect trip for the 4th to be skipped
      expect(new Date(trips[0].arrivalTime * 1000).getUTCDate()).toBe(5)
    })

    test("with interpolated stop_times", async () => {
      const trips = await getTripSchedule("testfeed:CITY,testfeed:NADAV")

      const interpolatedTrip = trips.find((t) => t.tripId === "testfeed:CITY2")

      expect(interpolatedTrip).toBeDefined()

      const arrival = new Date(interpolatedTrip!.arrivalTime * 1000)
      expect(arrival.getUTCHours()).toBe(15)
      expect(arrival.getUTCMinutes()).toBe(42)
      expect(arrival.getUTCSeconds()).toBe(0)

      expect(interpolatedTrip!.arrivalTime).toBe(
        interpolatedTrip!.departureTime,
      )
    })

    test("with frequency-based trip", async () => {
      const trips = await getTripSchedule("testfeed:AB,testfeed:BEATTY_AIRPORT")

      // We want to skip frequency-based trips for now since they are unsupported
      expect(trips.some((trip) => trip.tripId === "testfeed:AB1")).toBe(false)
    })

    describe("with GTFS-RT updates", () => {
      afterEach(() => {
        fakeGtfs.setTripUpdates([])
        fakeGtfs.setSimulateTripUpdatesFailure(false)
      })

      test("falls back to static schedule if GTFS-RT fails", async () => {
        fakeGtfs.setSimulateTripUpdatesFailure(true)

        const trips = await getTripSchedule()
        expect(trips.length).toBeGreaterThan(0)
        expect(trips).toMatchSnapshot()
      })

      test("with same trip on multiple days", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA",
              startDate: "20080104",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                arrival: {
                  time: 1199455200,
                },
              },
            ],
          },
          {
            trip: {
              tripId: "STBA",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                arrival: {
                  time: 1199541600,
                },
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        const updatedTrips = trips.filter(
          (trip) => trip.tripId === "testfeed:STBA",
        )

        expect(updatedTrips).toHaveLength(2)

        expect(updatedTrips[0].arrivalTime).toBe(1199455200)
        expect(updatedTrips[0].isRealtime).toBe(true)

        expect(updatedTrips[1].arrivalTime).toBe(1199541600)
        expect(updatedTrips[1].isRealtime).toBe(true)
      })

      test("with same trip on multiple days using ambiguous start_date", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                arrival: {
                  time: 1199455200,
                },
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        const updatedTrips = trips.filter(
          (trip) => trip.tripId === "testfeed:STBA",
        )

        expect(updatedTrips).toHaveLength(2)

        expect(updatedTrips[0].arrivalTime).toBe(1199455200)
        expect(updatedTrips[0].isRealtime).toBe(true)

        expect(updatedTrips[1].arrivalTime).toBe(1199541600)
        expect(updatedTrips[1].isRealtime).toBe(false)
      })

      test("with cancelled trip on multiple days using ambiguous start_date", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.CANCELED,
            },
          },
        ])

        const trips = await getTripSchedule()
        const remainingUncancelledTrips = trips.filter(
          (trip) => trip.tripId === "testfeed:STBA",
        )

        // Expect that we have only cancelled one of the two STBA trips
        expect(remainingUncancelledTrips).toHaveLength(1)
        expect(remainingUncancelledTrips[0].arrivalTime).toBe(1199541600)
      })

      test("with skipped stop on multiple days using ambiguous start_date", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                scheduleRelationship:
                  GtfsRt.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED,
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        const remainingUncancelledTrips = trips.filter(
          (trip) => trip.tripId === "testfeed:STBA",
        )

        // Expect that we have only cancelled one of the two STBA trips
        expect(remainingUncancelledTrips).toHaveLength(1)
        expect(remainingUncancelledTrips[0].arrivalTime).toBe(1199541600)
      })

      test("with time update more than 90 minutes deviated from schedule", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA",
              startDate: "20080104",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                arrival: {
                  time: 1199460700,
                },
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        expect(
          trips.some(
            (trip) =>
              trip.tripId === "testfeed:STBA" &&
              trip.arrivalTime === 1199460700,
          ),
        ).toBe(false)
        expect(
          trips.some(
            (trip) => trip.tripId === "testfeed:STBA" && trip.isRealtime,
          ),
        ).toBe(false)
      })

      test.each(["arrival", "departure"])(
        "with %s time update",
        async (arrivalOrDeparture: string) => {
          fakeGtfs.setTripUpdates([
            {
              trip: {
                tripId: "STBA",
                startDate: "20080104",
                scheduleRelationship:
                  GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
              },
              stopTimeUpdate: [
                {
                  stopId: "STAGECOACH",
                  [arrivalOrDeparture]: {
                    time: 1199455230,
                  },
                },
              ],
            },
          ])

          const trips = await getTripSchedule()
          const trip = trips.find((trip) => trip.tripId === "testfeed:STBA")

          expect(trip).toBeDefined()
          expect(trip!.arrivalTime).toBe(1199455230)
          expect(trip!.departureTime).toBe(1199455230)
          expect(trip!.isRealtime).toBe(true)
        },
      )

      test.each(["arrival", "departure"])(
        "with %s delay",
        async (arrivalOrDeparture: string) => {
          fakeGtfs.setTripUpdates([
            {
              trip: {
                tripId: "STBA",
                startDate: "20080104",
                scheduleRelationship:
                  GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
              },
              stopTimeUpdate: [
                {
                  stopId: "STAGECOACH",
                  [arrivalOrDeparture]: {
                    delay: 30,
                  },
                },
              ],
            },
          ])

          const trips = await getTripSchedule()
          const trip = trips.find((trip) => trip.tripId === "testfeed:STBA")

          expect(trip).toBeDefined()
          expect(trip!.arrivalTime).toBe(1199455230)
          expect(trip!.departureTime).toBe(1199455230)
          expect(trip!.isRealtime).toBe(true)
        },
      )

      test("with cancelled trip", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "AAMV1",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.CANCELED,
            },
          },
        ])

        const trips = await getTripSchedule()
        expect(trips.some((trip) => trip.tripId === "testfeed:AAMV1")).toBe(
          false,
        )
      })

      test("with update to overnight trip (crossing midnight)", async () => {
        dateSpy.mockImplementation(() =>
          new Date("2008-01-05T08:25:00.000Z").getTime(),
        )

        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "STBA_OVERNIGHT",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "STAGECOACH",
                arrival: {
                  time: 1199521830,
                },
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        const overnightTrips = trips.filter(
          (trip) => trip.tripId === "testfeed:STBA_OVERNIGHT",
        )

        expect(overnightTrips).toHaveLength(2)

        expect(overnightTrips[0].arrivalTime).toBe(1199521830)
        expect(overnightTrips[0].departureTime).toBe(1199521830)
        expect(overnightTrips[0].isRealtime).toBe(true)

        expect(overnightTrips[1].arrivalTime).toBe(1199608200)
        expect(overnightTrips[1].departureTime).toBe(1199608200)
        expect(overnightTrips[1].isRealtime).toBe(false)
      })

      test("with skipped stop by stop_id", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "AAMV2",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "BEATTY_AIRPORT",
                scheduleRelationship:
                  GtfsRt.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED,
              },
            ],
          },
          {
            trip: {
              tripId: "AAMV3",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopId: "SOME_OTHER_STOP",
                scheduleRelationship:
                  GtfsRt.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED,
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        expect(trips.some((trip) => trip.tripId === "testfeed:AAMV2")).toBe(
          false,
        )

        expect(trips.some((trip) => trip.tripId === "testfeed:AAMV3")).toBe(
          true,
        )
      })

      test("with skipped stop by stop_sequence", async () => {
        fakeGtfs.setTripUpdates([
          {
            trip: {
              tripId: "AAMV2",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopSequence: 2, // stop_id: BEATTY_AIRPORT
                scheduleRelationship:
                  GtfsRt.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED,
              },
            ],
          },
          {
            trip: {
              tripId: "AAMV3",
              startDate: "20080105",
              scheduleRelationship:
                GtfsRt.TripDescriptor.ScheduleRelationship.SCHEDULED,
            },
            stopTimeUpdate: [
              {
                stopSequence: 2, // stop_id: AMV
                scheduleRelationship:
                  GtfsRt.TripUpdate.StopTimeUpdate.ScheduleRelationship.SKIPPED,
              },
            ],
          },
        ])

        const trips = await getTripSchedule()
        expect(trips.some((trip) => trip.tripId === "testfeed:AAMV2")).toBe(
          false,
        )

        expect(trips.some((trip) => trip.tripId === "testfeed:AAMV3")).toBe(
          true,
        )
      })
    })
  })
})
