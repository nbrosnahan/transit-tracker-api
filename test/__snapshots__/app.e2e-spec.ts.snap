// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E test GET /feeds/service-areas 1`] = `
{
  "features": [
    {
      "geometry": {
        "coordinates": [
          [
            [
              -116.40094,
              36.641495,
            ],
            [
              -117.13316,
              36.42529,
            ],
            [
              -116.81797,
              36.88108,
            ],
            [
              -116.76821,
              36.914894,
            ],
            [
              -116.75168,
              36.915684,
            ],
            [
              -116.40094,
              36.641495,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": {},
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`E2E test GET /schedule/:routeStopPairs with GTFS-RT updates falls back to static schedule if GTFS-RT fails 1`] = `
[
  {
    "arrivalTime": 1199455200,
    "departureTime": 1199455200,
    "headsign": "Shuttle",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA",
  },
  {
    "arrivalTime": 1199521800,
    "departureTime": 1199521800,
    "headsign": "Shuttle Overnight",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA_OVERNIGHT",
  },
  {
    "arrivalTime": 1199541600,
    "departureTime": 1199541600,
    "headsign": "Shuttle",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA",
  },
  {
    "arrivalTime": 1199548800,
    "departureTime": 1199548800,
    "headsign": "to Amargosa Valley",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV1",
  },
  {
    "arrivalTime": 1199559600,
    "departureTime": 1199559600,
    "headsign": "to Airport",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV2",
  },
  {
    "arrivalTime": 1199566800,
    "departureTime": 1199566800,
    "headsign": "to Amargosa Valley",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV3",
  },
  {
    "arrivalTime": 1199577600,
    "departureTime": 1199577600,
    "headsign": "to Airport",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV4",
  },
  {
    "arrivalTime": 1199608200,
    "departureTime": 1199608200,
    "headsign": "Shuttle Overnight",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA_OVERNIGHT",
  },
]
`;

exports[`E2E test GET /schedule/:routeStopPairs with service exception in static schedule 1`] = `
[
  {
    "arrivalTime": 1181048400,
    "departureTime": 1181048400,
    "headsign": "Shuttle",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA",
  },
  {
    "arrivalTime": 1181115000,
    "departureTime": 1181115000,
    "headsign": "Shuttle Overnight",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA_OVERNIGHT",
  },
]
`;

exports[`E2E test GET /schedule/:routeStopPairs with static schedule 1`] = `
[
  {
    "arrivalTime": 1199455200,
    "departureTime": 1199455200,
    "headsign": "Shuttle",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA",
  },
  {
    "arrivalTime": 1199521800,
    "departureTime": 1199521800,
    "headsign": "Shuttle Overnight",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA_OVERNIGHT",
  },
  {
    "arrivalTime": 1199541600,
    "departureTime": 1199541600,
    "headsign": "Shuttle",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA",
  },
  {
    "arrivalTime": 1199548800,
    "departureTime": 1199548800,
    "headsign": "to Amargosa Valley",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV1",
  },
  {
    "arrivalTime": 1199559600,
    "departureTime": 1199559600,
    "headsign": "to Airport",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV2",
  },
  {
    "arrivalTime": 1199566800,
    "departureTime": 1199566800,
    "headsign": "to Amargosa Valley",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV3",
  },
  {
    "arrivalTime": 1199577600,
    "departureTime": 1199577600,
    "headsign": "to Airport",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:AAMV",
    "routeName": "50",
    "stopId": "testfeed:BEATTY_AIRPORT",
    "stopName": "Nye County Airport (Demo)",
    "tripId": "testfeed:AAMV4",
  },
  {
    "arrivalTime": 1199608200,
    "departureTime": 1199608200,
    "headsign": "Shuttle Overnight",
    "isRealtime": false,
    "routeColor": null,
    "routeId": "testfeed:STBA",
    "routeName": "30",
    "stopId": "testfeed:STAGECOACH",
    "stopName": "Stagecoach Hotel & Casino (Demo)",
    "tripId": "testfeed:STBA_OVERNIGHT",
  },
]
`;

exports[`E2E test GET /stops/:id/routes 1`] = `
[
  {
    "color": null,
    "headsigns": [
      "to Airport",
      "to Amargosa Valley",
    ],
    "name": "50",
    "routeId": "testfeed:AAMV",
  },
]
`;

exports[`E2E test GET /stops/within/:bbox 1`] = `
[
  {
    "lat": 36.914894,
    "lon": -116.76821,
    "name": "North Ave / D Ave N (Demo)",
    "stopCode": null,
    "stopId": "testfeed:NADAV",
  },
  {
    "lat": 36.914944,
    "lon": -116.761475,
    "name": "North Ave / N A Ave (Demo)",
    "stopCode": null,
    "stopId": "testfeed:NANAA",
  },
]
`;
